ethcore/src/engines/clique/mod.rs

```diff
@@ -234,111 +236,174 @@ impl Clique {
 
 		state.calc_next_timestamp(header, self.period);
 
 		Ok(state)
 	}
 
 	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		// FIXME: Remove??
 		self.block_state_by_hash.write().get_mut(hash).cloned()
 	}
 
 	/// Get `CliqueBlockState` for given header, backfill from last checkpoint if needed.
 	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		// TODO: Split and simplify this function. Allow retrieving by hash / block_number (keep private)
+
+		let block_number = header.number();
+		let block_hash = header.hash();
+
+		//////
+
 		let mut block_state_by_hash = self.block_state_by_hash.write();
-		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+		if let Some(state) = block_state_by_hash.get_mut(&block_hash) {
 			return Ok(state.clone());
 		}
+
 		// If we are looking for an checkpoint block state, we can directly reconstruct it.
-		if header.number() % self.epoch_length == 0 {
+		if block_number % self.epoch_length == 0 {
 			let state = self.new_checkpoint_state(header)?;
-			block_state_by_hash.insert(header.hash(), state.clone());
+			block_state_by_hash.insert(block_hash, state.clone());
 			return Ok(state);
 		}
+
 		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
-		match self.client.read().as_ref().and_then(|w| w.upgrade()) {
-			None => {
-				return Err(EngineError::RequiresClient)?;
+		let client = self.client.read().as_ref().and_then(|w| w.upgrade()).ok_or_else(|| EngineError::RequiresClient)?;
+
+		let last_checkpoint_number = block_number - block_number % self.epoch_length as u64;
+		debug_assert_ne!(last_checkpoint_number, block_number);
+		let capacity = block_number - last_checkpoint_number + 1;
+
+		let chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity(capacity as usize);
+
+		// Put ourselves in.
+		chain.push_front(header.clone());
+
+		// populate chain to last checkpoint
+		loop {
+			let (last_parent_hash, last_num) = {
+				let l = chain.front().expect("chain has at least one element; qed");
+				(*l.parent_hash(), l.number())
+			};
+
+			if last_num == last_checkpoint_number + 1 {
+				break;
 			}
-			Some(c) => {
-				let last_checkpoint_number = header.number() - header.number() % self.epoch_length as u64;
-				debug_assert_ne!(last_checkpoint_number, header.number());
 
-				let mut chain: &mut VecDeque<Header> = &mut VecDeque::with_capacity(
-					(header.number() - last_checkpoint_number + 1) as usize);
+			let query = BlockId::Hash(last_parent_hash);
+			let next_header = client.block_header(query).ok_or_else(|| BlockError::UnknownParent(last_parent_hash))?;
+			chain.push_front(next_header.decode()?);
+		}
 
-				// Put ourselves in.
-				chain.push_front(header.clone());
+		// Catching up state, note that we don't really store block state for intermediary blocks,
+		// for speed.
+		let backfill_start = time::Instant::now();
```

Comment:
I think it makes sense to start the back-filling profiling before the loop above.  I think that is where the biggest performance hit (Disk IO) from back-filling is.

Reply:
I tried to avoid changing any previous functionality here.
The diff looks a bit confusing. I tried to reduce how nested the code by early returning instead of nested match/none/some.
I would say this is a good example of a change that I could move to another PR to simplify the review a bit.

Reply:
also there is a mixup of tabs/spaces that complicates things a bit too.
the `.editorconfig` is set to tabs so maybe we can clean this up a bit too.

Reply:
I'm going to open a PR

Reply:
hopefully we can coordinate so merging is simple. Otherwise, we will up reworking the same code multiple time.

Reply:
It's a three line change.  I won't be making any deep refactors in parallel to this PR.